src/Core/BaseSettings.h

```diff
@@ -0,0 +1,833 @@
+
+    private:
+        friend class BaseSettings;
+        SettingFieldRef(const typename Traits::Data & data_, const typename Traits::Accessor & accessor_, size_t index_) : data(&data_), accessor(&accessor_), index(index_) {}
+        SettingFieldRef(const CustomSettingMap::mapped_type & custom_setting_);
+
+        const typename Traits::Data * data = nullptr;
+        const typename Traits::Accessor * accessor = nullptr;
+        size_t index = 0;
+        std::conditional_t<Traits::allow_custom_settings, const CustomSettingMap::mapped_type*, boost::blank> custom_setting = {};
+    };
+
+    class Iterator
+    {
+    public:
+        Iterator & operator++();
+        Iterator operator++(int);
+        SettingFieldRef operator *() const;
+
+        bool operator ==(const Iterator & other) const;
+        bool operator !=(const Iterator & other) const { return !(*this == other); }
+
+    private:
+        friend class BaseSettings;
+        Iterator(const BaseSettings & settings_, const typename Traits::Accessor & accessor_, bool skip_changed_, bool skip_unchanged_);
+        void doSkip();
+
+        const BaseSettings * settings = nullptr;
+        const typename Traits::Accessor * accessor = nullptr;
+        size_t index = 0;
+        std::conditional_t<Traits::allow_custom_settings, CustomSettingMap::const_iterator, boost::blank> custom_settings_iterator;
+        bool skip_changed = false;
+        bool skip_unchanged = false;
+    };
+
+    class Range
+    {
+    public:
+        Range(const BaseSettings & settings_, bool including_changed_, bool including_unchanged_) : settings(settings_), accessor(Traits::Accessor::instance()), including_changed(including_changed_), including_unchanged(including_unchanged_) {}
+        Iterator begin() const { return Iterator(settings, accessor, !including_changed, !including_unchanged); }
+        Iterator end() const { return Iterator(settings, accessor, true, true); }
+
+    private:
+        const BaseSettings & settings;
+        const typename Traits::Accessor & accessor;
+        bool including_changed = true;
+        bool including_unchanged = true;
+    };
+
+    Range all(bool including_changed = true, bool including_unchanged = true) const { return Range{*this, including_changed, including_unchanged}; }
+    Range allChanged() const { return all(true, false); }
+    Range allUnchanged() const { return all(false, true); }
+
+    Iterator begin() const { return allChanged().begin(); }
+    Iterator end() const { return allChanged().end(); }
+
+private:
+    SettingFieldCustom & getCustomSetting(const std::string_view & name);
+    const SettingFieldCustom & getCustomSetting(const std::string_view & name) const;
+    const SettingFieldCustom * tryGetCustomSetting(const std::string_view & name) const;
+
+    std::conditional_t<Traits::allow_custom_settings, CustomSettingMap, boost::blank> custom_settings_map;
```

Comment:
As far as I see, there is no way to iterate only over custom settings. Maybe we can introduce 

```
Range allCustom() const;
```

helper? That would be convenient.

Reply:
done